statsfs/otelstats/subsys_metric_parser.go

```diff
@@ -0,0 +1,199 @@
+// SubsysMetric stuct
+// Example:
+//	Input:
+//		m.SubSystemPath = /sys/kernel/stats/net
+// 		m.SubSystemName = net
+//		m.StatsfsPath = /sys/kernel/stats
+//		path = /sys/kernel/stats/net/eth0/latency
+//	Output:
+//		metricName = net/latency
+//		label = /eth0
+//		new entry in m.Metrics[net/latency]:
+//			MetricInfo{Label: /eth0, Path: /sys/kernel/stats/net/eth0/latency}
+func (m SubsysMetrics) updateMetricMapOneEntry(path string) {
+	metricName := m.getMetricName(path)
+	metricInfo := m.getMetricInfo(path)
+	m.Metrics[metricName] = append(m.Metrics[metricName], metricInfo)
+}
+
+func (m SubsysMetrics) updateMetricMap(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		log.Printf("Failed to walk to file %v\n", path)
+	}
+
+	if !info.IsDir() {
+		m.updateMetricMapOneEntry(path)
+	}
+	return err
+}
+
+func (m SubsysMetrics) constructMetricMap() {
+	err := filepath.Walk(m.SubSystemPath, m.updateMetricMap)
+	if err != nil {
+		log.Printf("Failed to parse metrics for subsystem %v at %v", m.SubSystemName, m.SubSystemPath)
+	}
+}
+
+func (m SubsysMetrics) print() {
+	fmt.Printf("StatsfsPath: %v\n", m.StatsfsPath)
+	fmt.Printf("SubSystemName: %v\n", m.SubSystemName)
+	fmt.Printf("SubSystemPath: %v\n", m.SubSystemPath)
+	fmt.Println("Metrics:")
+	for metricName, labels := range m.Metrics {
+		fmt.Printf("\tmetricName: %v,\n\tinfo: \n", metricName)
+		for _, label := range labels {
+			fmt.Printf("\t\tLabel: %v, Path: %v\n", label.Label, label.Path)
+		}
+	}
+}
+
+// Print prints StatsfsMetrics struct
+func (m StatsfsMetrics) Print() {
+	fmt.Print("\n####################################\n")
+	fmt.Printf("StatsfsPath: %v\n\n", m.StatsfsPath)
+	for subsysName, subsysMetrics := range m.Metrics {
+		fmt.Println("------------------")
+		fmt.Printf("Statsfs metrics for subsystem %v:\n\n", subsysName)
+		subsysMetrics.print()
+	}
+	fmt.Printf("####################################\n\n")
+}
+
+// MetricInfo contains a Label used to identify the specific device
+// and a Path from where the metric for this device could be retrieved
+type MetricInfo struct {
+	Label string
+	Path  string
+}
+
+// SubsysMetrics is a struct that represents metrics of a subsystem.
+// Path gives the base path to the subsystem's stats in statsfs
+// (usually /sys/kernel/stats/subsystemName), Metrics is a map with key
+// being the metric name, and value being a list of labels (devices with
+// the metric registered)
+type SubsysMetrics struct {
+	StatsfsPath   string
+	SubSystemName string
+	SubSystemPath string
+	Metrics       map[string][]MetricInfo
+}
+
+// StatsfsMetrics is a struct that represents metrics available in a statsfs
+// filesystem found at StatsfsPath
+// Each subsystem metrics is represented with a SubsysMetrics struct
+type StatsfsMetrics struct {
+	StatsfsPath string
+	Metrics     map[string]SubsysMetrics
+}
+
+func initSubsysMetricStruct(statsfsPath, subsystemName string) SubsysMetrics {
```

Comment:
In golang, we usually call it

`newSubsysMetric(...)`

Also, please placed it just under the `type SubsysMetrics struct`.